File: src/lib/cli/incidents.ts
/** biome-ignore-all lint/suspicious/noConsole: This prints the incidents, it's part of the CLI */
import { loadConfig } from "../../config";
import { initDb } from "../db";
import { formatDate } from "../helpers";
import { IncidentStore } from "../incident-store";
import { logger } from "../logger";

type ListStore = Pick<IncidentStore, "listIncidents">;
type GetStore = Pick<IncidentStore, "getIncident">;

// Each CLI invocation is a fresh process — unlike `baba start`, nothing else
// has called initDb() yet, so IncidentStore's shared getDb() would throw.
async function defaultStore(configPath?: string): Promise<IncidentStore> {
	await loadConfig(configPath);
	initDb();
	return new IncidentStore();
}

export async function listIncidents(
	opts: { limit: string; config?: string },
	store?: ListStore,
) {
	const resolvedStore = store ?? (await defaultStore(opts.config));
	const incidents = resolvedStore.listIncidents(Number(opts.limit));

	if (incidents.length === 0) {
		logger.info("No incidents recorded.");
		return;
	}

	const col = (s: string, w: number) => s.padEnd(w).slice(0, w);
	const header =
		`${"ID".padEnd(6)}${"Metric".padEnd(28)}${"Volume".padEnd(16)}` +
		`${"Started".padEnd(22)}${"Status".padEnd(10)}${"Peak".padEnd(10)}` +
		`${"Threshold".padEnd(10)}${"Notifs".padEnd(6)}`;
	const divider = "─".repeat(header.length);

	console.log(header);
	console.log(divider);

	for (const inc of incidents) {
		const status = inc.resolved_at ? "RESOLVED" : "OPEN";
		console.log(
			`${col(String(inc.id), 6)}` +
				`${col(inc.metric, 28)}` +
				`${col(inc.volume ?? "-", 16)}` +
				`${col(formatDate(inc.started_at), 22)}` +
				`${col(status, 10)}` +
				`${col(String(inc.peak_value), 10)}` +
				`${col(String(inc.threshold), 10)}` +
				`${col(String(inc.notification_count), 6)}`,
		);
	}
}

interface GetIncidentProps {
	store?: GetStore;
	exit: (code: number) => void;
	config?: string;
}

export async function getIncident(
	id: string,
	props: GetIncidentProps = { exit: process.exit },
) {
	const store = props.store ?? (await defaultStore(props.config));
	const incident = store.getIncident(Number(id));

	if (!incident) {
		logger.error(`Incident #${id} not found.`);
		props.exit(1);
		return;
	}

	const status = incident.resolved_at ? "RESOLVED" : "OPEN";
	console.log(`\nIncident #${incident.id}`);
	console.log(`  Metric:     ${incident.metric}`);
	if (incident.volume) {
		console.log(`  Volume:     ${incident.volume}`);
	}
	console.log(`  Status:     ${status}`);
	console.log(`  Started:    ${formatDate(incident.started_at)}`);
	if (incident.resolved_at) {
		console.log(`  Resolved:   ${formatDate(incident.resolved_at)}`);
	}
	console.log(`  Peak value: ${incident.peak_value}`);
	console.log(`  Threshold:  ${incident.threshold}`);

	if (incident.notifications.length === 0) {
		console.log("\n  No notifications recorded.");
	} else {
		console.log(`\n  Notifications (${incident.notifications.length}):`);
		for (const n of incident.notifications) {
			const ok = n.succeeded ? "✓" : "✗";
			console.log(`    ${formatDate(n.sent_at)}  ${n.type.padEnd(10)} ${ok}`);
		}
	}
	console.log();
}

File: cmd/baba/incidents.go
package main

import (
	"cmp"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"

	"github.com/orochibraru/baba/internal/incidents"
)

func openStore(configPath string) (*incidents.Store, error) {
	c, err := load(configPath)
	if err != nil {
		return nil, err
	}
	return incidents.Open(c.Database.Path)
}

func status(i *incidents.Incident) string {
	if i.ResolvedAt.IsZero() {
		return "OPEN"
	}
	return "RESOLVED"
}

func date(t time.Time) string { return t.UTC().Format(time.DateTime) }

func listIncidents(w io.Writer, configPath string, limit int) error {
	store, err := openStore(configPath)
	if err != nil {
		return err
	}
	list := store.List(limit)
	if len(list) == 0 {
		fmt.Fprintln(w, "No incidents recorded.")
		return nil
	}
	widths := []int{6, 28, 16, 22, 10, 10, 10, 6}
	row := func(cells ...string) {
		var b strings.Builder
		for i, cell := range cells {
			fmt.Fprintf(&b, "%-*s", widths[i], cell[:min(len(cell), widths[i])])
		}
		fmt.Fprintln(w, strings.TrimRight(b.String(), " "))
	}
	row("ID", "Metric", "Volume", "Started", "Status", "Peak", "Threshold", "Notifs")
	fmt.Fprintln(w, strings.Repeat("─", 108))
	for _, i := range list {
		row(strconv.Itoa(i.ID), i.Metric, cmp.Or(i.Volume, "-"), date(i.StartedAt), status(i),
			strconv.FormatFloat(i.PeakValue, 'f', -1, 64), strconv.FormatFloat(i.Threshold, 'f', -1, 64), strconv.Itoa(len(i.Notifications)))
	}
	return nil
}

func getIncident(w io.Writer, configPath string, id int) error {
	store, err := openStore(configPath)
	if err != nil {
		return err
	}
	i := store.Get(id)
	if i == nil {
		return fmt.Errorf("incident #%d not found", id)
	}
	fmt.Fprintf(w, "\nIncident #%d\n  Metric:     %s\n", i.ID, i.Metric)
	if i.Volume != "" {
		fmt.Fprintf(w, "  Volume:     %s\n", i.Volume)
	}
	fmt.Fprintf(w, "  Status:     %s\n  Started:    %s\n", status(i), date(i.StartedAt))
	if !i.ResolvedAt.IsZero() {
		fmt.Fprintf(w, "  Resolved:   %s\n", date(i.ResolvedAt))
	}
	fmt.Fprintf(w, "  Peak value: %g\n  Threshold:  %g\n", i.PeakValue, i.Threshold)
	if len(i.Notifications) == 0 {
		fmt.Fprintln(w, "\n  No notifications recorded.")
	} else {
		fmt.Fprintf(w, "\n  Notifications (%d):\n", len(i.Notifications))
		for _, n := range i.Notifications {
			fmt.Fprintf(w, "    %s  %-10s %s\n", date(n.SentAt), n.Type, map[bool]string{true: "✓", false: "✗"}[n.Succeeded])
		}
	}
	fmt.Fprintln(w)
	return nil
}
